build.rs

```diff
@@ -145,6 +151,71 @@ fn generate_aarch64_outlined_atomics() {
     std::fs::write(dst, buf).unwrap();
 }
 
+/// Emit directives for features we expect to support that aren't in `Cargo.toml`.
+///
+/// These are mostly cfg elements emitted by this `build.rs`.
+fn configure_check_cfg() {
+    // Functions where we can set the "optimized-c" flag
+    const HAS_OPTIMIZED_C: &[&str] = &[
+        "__ashldi3",
+        "__ashlsi3",
+        "__ashrdi3",
+        "__ashrsi3",
+        "__clzsi2",
+        "__divdi3",
+        "__divsi3",
+        "__divmoddi4",
+        "__divmodsi4",
+        "__divmodsi4",
+        "__divmodti4",
+        "__lshrdi3",
+        "__lshrsi3",
+        "__moddi3",
+        "__modsi3",
+        "__muldi3",
+        "__udivdi3",
+        "__udivmoddi4",
+        "__udivmodsi4",
+        "__udivsi3",
+        "__umoddi3",
+        "__umodsi3",
+    ];
+
+    // Build a list of all aarch64 atomic operation functions
+    let mut aarch_atomic = Vec::new();
+    for aarch_op in ["cas", "ldadd", "ldclr", "ldeor", "ldset", "swp"] {
+        for op_size in [1, 2, 4, 8].iter().chain(if aarch_op == "cas" {
+            [16].as_slice()
+        } else {
+            [].as_slice()
+        }) {
+            for ordering in ["relax", "acq", "rel", "acq_rel"] {
+                aarch_atomic.push(format!("__aarch64_{}{}_{}", aarch_op, op_size, ordering));
+            }
+        }
+    }
+
+    for fn_name in HAS_OPTIMIZED_C
+        .iter()
+        .copied()
+        .chain(aarch_atomic.iter().map(|s| s.as_str()))
+    {
+        println!(
+            "cargo::rustc-check-cfg=cfg({}, values(\"optimized-c\"))",
+            fn_name
+        );
+    }
+
+    // FIXME: this feature definitely exists in
+    // `rustc --print target-features --target sparc64-unknown-linux-gnu`. Why does check-cfg
+    // flag it?
+    println!("cargo::rustc-check-cfg=cfg(target_feature, values(\"vis3\"))");
```

Comment:
Probably because `rustc` (the front-end) doesn't know about any of the sparc target features. The full list of known target features is defined [here](https://github.com/rust-lang/rust/blob/d287f3e4eeaf680e8fe875f1ec75cca68f357d30/compiler/rustc_target/src/target_features.rs) and there isn't any entry for the Sparc architecture so any attempt to use them with the front-end will result in a warning.

It may still work since we pass them anyway to codegen but that's not guaranteed at all.

Reply:
Ah, makes sense. I just updated the note.